src/traits/snark.rs

```diff
@@ -61,6 +62,44 @@ pub trait RelaxedR1CSSNARKTrait<G: Group>:
   fn verify(&self, vk: &Self::VerifierKey, U: &RelaxedR1CSInstance<G>) -> Result<(), NovaError>;
 }
 
+/// A trait that defines the behavior of a `zkSNARK` to prove knowledge of satisfying witness to batches of relaxed R1CS instances.
+pub trait BatchedRelaxedR1CSSNARKTrait<G: Group>:
+  Send + Sync + Serialize + for<'de> Deserialize<'de>
+{
+  /// A type that represents the prover's key
+  type ProverKey: Send + Sync + Serialize + for<'de> Deserialize<'de>; // + Abomonation; //TODO: Uncomment this
+
+  /// A type that represents the verifier's key
+  type VerifierKey: Send + Sync + Serialize + for<'de> Deserialize<'de> + DigestHelperTrait<G>;
+  //+ Abomonation; // TODO: Uncomment this
+
+  /// This associated function (not a method) provides a hint that offers
+  /// a minimum sizing cue for the commitment key used by this SNARK
+  /// implementation. The commitment key passed in setup should then
+  /// be at least as large as this hint.
+  fn commitment_key_floor() -> Box<dyn for<'a> Fn(&'a R1CSShape<G>) -> usize> {
+    default_commitment_key_hint()
+  }
+
+  /// Produces the keys for the prover and the verifier
+  fn setup(
+    ck: &CommitmentKey<G>,
+    S: &[R1CSShape<G>],
+  ) -> Result<(Self::ProverKey, Self::VerifierKey), NovaError>;
+
+  /// Produces a new SNARK for a batch of relaxed R1CS
+  fn prove(
+    ck: &CommitmentKey<G>,
+    pk: &Self::ProverKey,
+    S: &[R1CSShape<G>],
+    U: &[RelaxedR1CSInstance<G>],
+    W: &[RelaxedR1CSWitness<G>],
```

Comment:
Can we make this take the inputs by value? Maybe even `impl IntoIterator<Item = ...>`?